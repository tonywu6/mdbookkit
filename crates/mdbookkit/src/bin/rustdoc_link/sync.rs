use std::{
    sync::{Arc, RwLock},
    task::Poll,
    time::Duration,
};

use anyhow::{bail, Result};
use tokio::{
    sync::{mpsc, Notify},
    task::JoinHandle,
    time,
};

pub struct EventSampling {
    pub buffer: Duration,
    pub timeout: Duration,
}

impl EventSampling {
    pub fn using<T>(self, rx: mpsc::Receiver<Poll<T>>) -> EventSampler<T>
    where
        T: Clone + Send + Sync + 'static,
    {
        EventSampler::new(rx, self)
    }
}

#[derive(Debug, Clone)]
pub struct EventSampler<T> {
    state: Arc<RwLock<State<T>>>,
    event: Arc<Notify>,
}

#[derive(Debug, Clone)]
enum State<T> {
    Pending,
    Ready(T),
    Timeout,
}

impl<T: Clone + Send + Sync + 'static> EventSampler<T> {
    fn new(
        mut rx: mpsc::Receiver<Poll<T>>,
        EventSampling { buffer, timeout }: EventSampling,
    ) -> Self {
        let state = Arc::new(RwLock::new(State::Pending));
        let event = Arc::new(Notify::new());

        tokio::spawn({
            let state = state.clone();
            let event = event.clone();
            async move {
                let mut abort: Option<JoinHandle<()>> = None;
                while let Some(value) = time::timeout(timeout, rx.recv()).await.transpose() {
                    if let Some(abort) = abort.take() {
                        abort.abort();
                    }
                    match value {
                        Ok(Poll::Ready(value)) => {
                            let event = event.clone();
                            let state = state.clone();
                            abort = Some(tokio::spawn(async move {
                                time::sleep(buffer).await;
                                *state.write().unwrap() = State::Ready(value);
                                event.notify_waiters();
                            }));
                        }
                        Ok(Poll::Pending) => {
                            *state.write().unwrap() = State::Pending;
                            event.notify_waiters();
                        }
                        Err(_) => {
                            *state.write().unwrap() = State::Timeout;
                            event.notify_waiters();
                        }
                    }
                }
            }
        });

        Self { event, state }
    }

    pub async fn wait(&self) -> Result<T> {
        loop {
            {
                match self.state.read().unwrap().clone() {
                    State::Pending => {}
                    State::Ready(value) => return Ok(value),
                    State::Timeout => bail!("timed out waiting for ready event"),
                }
            }
            self.event.notified().await;
        }
    }
}
