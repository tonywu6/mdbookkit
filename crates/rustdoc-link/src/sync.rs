use std::{
    sync::{Arc, RwLock},
    task::Poll,
    time::Duration,
};

use tokio::{
    sync::{mpsc, Notify},
    task::JoinHandle,
    time::sleep,
};

#[derive(Debug, Clone)]
pub struct Subsample<T> {
    state: Arc<RwLock<Poll<T>>>,
    event: Arc<Notify>,
}

impl<T: Clone + Send + Sync + 'static> Subsample<T> {
    pub fn new(mut rx: mpsc::Receiver<Poll<T>>, wait: Duration) -> Self {
        let state = Arc::new(RwLock::new(Poll::Pending));
        let event = Arc::new(Notify::new());

        tokio::spawn({
            let state = state.clone();
            let event = event.clone();
            async move {
                let mut abort: Option<JoinHandle<()>> = None;
                while let Some(value) = rx.recv().await {
                    if let Some(abort) = abort.take() {
                        abort.abort();
                    }
                    if value.is_ready() {
                        let event = event.clone();
                        let state = state.clone();
                        abort = Some(tokio::spawn(async move {
                            sleep(wait).await;
                            *state.write().unwrap() = value;
                            event.notify_waiters();
                        }));
                    } else {
                        *state.write().unwrap() = value;
                    }
                }
            }
        });

        Self { event, state }
    }

    pub async fn wait(&self) -> T {
        loop {
            {
                if let Poll::Ready(value) = self.state.read().unwrap().clone() {
                    return value;
                }
            }
            self.event.notified().await;
        }
    }
}
