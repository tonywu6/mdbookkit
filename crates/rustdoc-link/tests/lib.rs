use std::sync::Arc;

use anyhow::{bail, Result};

use console::set_colors_enabled_stderr;
use log::LevelFilter;
use mdbook_rustdoc_link::{
    env::{Config, Environment},
    logger::ConsoleLogger,
    Client, Pages, Resolver,
};
use similar::{ChangeTag, TextDiff};
use tap::Pipe;
use tokio::task::JoinSet;
use util_testing::{portable_snapshots, test_document, TestDocument};

async fn snapshot(client: Arc<Client>, TestDocument { source, name }: TestDocument) -> Result<()> {
    let stream = client.env().markdown(source).into_offset_iter();
    let mut page = Pages::one(source, stream)?;
    client.resolve(&mut page).await?;
    let output = page.get(&client.env().emit_config())?.to_string();

    portable_snapshots!().test(|| insta::assert_snapshot!(name.clone(), output))?;
    assert_no_whitespace_change(source, &output)?;

    let report = page
        .reporter()
        .level(LevelFilter::Trace)
        .paths(|_| name.clone())
        .report()
        .unwrap();
    portable_snapshots!().test(|| insta::assert_snapshot!(format!("{name}.stderr"), report))?;

    Ok(())
}

fn assert_no_whitespace_change(source: &str, output: &str) -> Result<()> {
    let changed_lines = TextDiff::from_words(source, output)
        .iter_all_changes()
        .filter_map(|change| {
            if matches!(change.tag(), ChangeTag::Equal) {
                return None;
            }
            if change.value().contains('\n') {
                Some(change.value())
            } else {
                None
            }
        })
        .collect::<Vec<_>>();

    if !changed_lines.is_empty() {
        bail!("unexpected whitespace change: {changed_lines:?}")
    } else {
        Ok(())
    }
}

#[tokio::test]
async fn test_snapshots() -> Result<()> {
    let client = setup()?;

    let tests = [
        test_document!("../../../docs/src/rustdoc-link/supported-syntax.md"),
        test_document!("../../../docs/src/rustdoc-link.md"),
        test_document!("tests/ra-known-quirks.md"),
    ];

    let errors = tests
        .map(|test| snapshot(client.clone(), test))
        .into_iter()
        .collect::<JoinSet<_>>()
        .join_all()
        .await
        .into_iter()
        .filter_map(Result::err)
        .collect::<Vec<_>>();

    if !errors.is_empty() {
        let errors = errors
            .iter()
            .map(|e| format!("{e:?}"))
            .collect::<Vec<_>>()
            .join("\n");
        panic!("{errors}")
    }

    client.drop().await?;

    Ok(())
}

fn setup() -> Result<Arc<Client>> {
    set_colors_enabled_stderr(false);
    ConsoleLogger::install();
    Config {
        rust_analyzer: Some("cargo run --release --package util-rust-analyzer -- analyzer".into()),
        ..Default::default()
    }
    .pipe(Environment::new)?
    .pipe(Client::new)
    .pipe(Arc::new)
    .pipe(Ok)
}
