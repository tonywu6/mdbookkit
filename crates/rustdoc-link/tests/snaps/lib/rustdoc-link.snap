---
source: crates/rustdoc-link/tests/lib.rs
expression: output
---
# mdbook-rustdoc-link

**[_rustdoc_-style linking][rustdoc] for [mdBook]** (with the help of [rust-analyzer]).

You write:

```md
The [`option`][std::option] and [`result`][std::result] modules define optional and
error-handling types, [`Option<T>`] and [`Result<T, E>`]. The [`iter`][std::iter] module
defines Rust's iterator trait, [`Iterator`], which works with the `for` loop to access
collections. [^1]
```

You get:

<figure class="fig-text">

The [`option`](https://doc.rust-lang.org/stable/core/option/index.html "std::option") and [`result`](https://doc.rust-lang.org/stable/core/result/index.html "std::result") modules define optional and
error-handling types, [`Option<T>`](https://doc.rust-lang.org/stable/core/option/enum.Option.html "Option<T>") and [`Result<T, E>`](https://doc.rust-lang.org/stable/core/result/enum.Result.html "Result<T, E>"). The [`iter`](https://doc.rust-lang.org/stable/core/iter/index.html "std::iter") module
defines Rust's iterator trait, [`Iterator`](https://doc.rust-lang.org/stable/core/iter/traits/iterator/trait.Iterator.html "Iterator"), which works with the `for` loop to access
collections. [^1]

</figure>

## Getting started

You will need to:

1. Have [rust-analyzer]:

   - If you already use the [VS Code extension][ra-extension], this project will
     automatically use the server binary that comes with it, no extra setup is needed!
   - Otherwise, [install][ra-install] rust-analyzer (e.g. via `rustup`) and make sure
     it's on your `PATH`.

2. Install this project:

   ```
   cargo install mdbook-rustdoc-link
   ```

3. Configure your `book.toml` to use it as a [preprocessor]:

   ```toml
   [book]
   title = "My Book"

   [preprocessor.rustdoc-link]
   # mdBook will run `mdbook-rustdoc-link`
   after = ["links"]
   # recommended, so that it can see content from {{#include}} as well
   ```

Now, when you want to link to a Rust item, such as a type, a function, etc., simply use
its name in place of a URL, like this:

```md
Like [`std::thread::spawn`], [`tokio::task::spawn`] returns a
[`JoinHandle`][tokio::task::JoinHandle] struct.
```

`mdbook-rustdoc-link` will then turn them into hyperlinks:

<figure class="fig-text">

Like [`std::thread::spawn`](https://doc.rust-lang.org/stable/std/thread/fn.spawn.html "std::thread::spawn"), [`tokio::task::spawn`](https://docs.rs/tokio/1.44.1/tokio/task/spawn/fn.spawn.html "tokio::task::spawn") returns a
[`JoinHandle`](https://docs.rs/tokio/1.44.1/tokio/runtime/task/join/struct.JoinHandle.html "tokio::task::JoinHandle") struct.

</figure>

This works in both `mdbook build` and `mdbook serve`!

![screen recording of mdbook-rustdoc-link during mdbook build](rustdoc-link/media/screencap.webp)

> [!IMPORTANT]
>
> It is assumed that you are running `mdbook` within a Cargo project.
>
> If you are working on a crate, and your book directory is within your source tree,
> such as next to `Cargo.toml`, then running `mdbook` from there will "just work".
>
> If your book doesn't belong to a Cargo project, refer to
> [Workspace layout](rustdoc-link/workspace-layout.md) for more information on how you
> can setup up the preprocessor.

> [!TIP]
>
> `mdbook-rustdoc-link` makes use of rust-analyzer's ["Open Docs"][open-docs] feature,
> which resolves links to documentation given a symbol.
>
> Items from `std` will generate links to <https://doc.rust-lang.org>, while items from
> third-party crates will generate links to <https://docs.rs>.
>
> So really, rust-analyzer is doing the heavy-lifting here. This crate is just the glue
> code :)

## Overview

## License

---

[^1]: Text adapted from [<cite>A Tour of The Rust Standard Library</cite>][tour]

<!-- prettier-ignore-start -->

[mdBook]: https://rust-lang.github.io/mdBook/
[open-docs]: https://rust-analyzer.github.io/book/features.html#open-docs
[preprocessor]: https://rust-lang.github.io/mdBook/format/configuration/preprocessors.html
[ra-extension]: https://marketplace.visualstudio.com/items?itemName=rust-lang.rust-analyzer
[ra-install]: https://rust-analyzer.github.io/book/rust_analyzer_binary.html
[rust-analyzer]: https://rust-analyzer.github.io/
[rustdoc]: https://doc.rust-lang.org/rustdoc/write-documentation/linking-to-items-by-name.html
[tour]: https://doc.rust-lang.org/stable/std/#a-tour-of-the-rust-standard-library

<!-- prettier-ignore-end -->
